Client/Frontend/Browser/TabManager.swift

```diff
@@ -378,66 +378,96 @@ class TabManager : NSObject {
-            self.title = coder.decodeObjectForKey("title") as? String
-            self.isPrivate = coder.decodeBoolForKey("isPrivate")
+        if let sessionDataInfo = self.sessionData?.jsonDictionary {
+            json["sessionData"] = sessionDataInfo
         }
 
-        func encodeWithCoder(coder: NSCoder) {
-            coder.encodeObject(sessionData, forKey: "sessionData")
-            coder.encodeObject(screenshotUUID, forKey: "screenshotUUID")
-            coder.encodeBool(isSelected, forKey: "isSelected")
-            coder.encodeObject(title, forKey: "title")
-            coder.encodeBool(isPrivate, forKey: "isPrivate")
+        return json
+    }
+
+    init?(browser: Browser, isSelected: Bool) {
+        self.screenshotUUID = browser.screenshotUUID
+        self.isSelected = isSelected
+        self.title = browser.displayTitle
+        self.isPrivate = browser.isPrivate
+        super.init()
+
+        if browser.sessionData == nil {
+            let currentItem: WKBackForwardListItem! = browser.webView?.backForwardList.currentItem
+
+            // Freshly created web views won't have any history entries at all.
+            // If we have no history, abort.
+            if currentItem == nil {
+                return nil
+            }
+
+            let backList = browser.webView?.backForwardList.backList ?? []
+            let forwardList = browser.webView?.backForwardList.forwardList ?? []
+            let urls = (backList + [currentItem] + forwardList).map { $0.URL }
+            let currentPage = -forwardList.count
+            self.sessionData = SessionData(currentPage: currentPage, urls: urls, lastUsedTime: browser.lastExecutedTime ?? NSDate.now())
+        } else {
+            self.sessionData = browser.sessionData
         }
     }
 
-    private func tabsStateArchivePath() -> String {
+    required init?(coder: NSCoder) {
+        self.sessionData = coder.decodeObjectForKey("sessionData") as? SessionData
+        self.screenshotUUID = coder.decodeObjectForKey("screenshotUUID") as? NSUUID
+        self.isSelected = coder.decodeBoolForKey("isSelected")
+        self.title = coder.decodeObjectForKey("title") as? String
+        self.isPrivate = coder.decodeBoolForKey("isPrivate")
+    }
+
+    func encodeWithCoder(coder: NSCoder) {
+        coder.encodeObject(sessionData, forKey: "sessionData")
+        coder.encodeObject(screenshotUUID, forKey: "screenshotUUID")
+        coder.encodeBool(isSelected, forKey: "isSelected")
+        coder.encodeObject(title, forKey: "title")
+        coder.encodeBool(isPrivate, forKey: "isPrivate")
+    }
+}
+
+extension TabManager {
+
+    static private func tabsStateArchivePath() -> String {
         let documentsPath = NSSearchPathForDirectoriesInDomains(.DocumentDirectory, .UserDomainMask, true)[0]
         return NSURL(fileURLWithPath: documentsPath).URLByAppendingPathComponent("tabsState.archive").path!
     }
 
+    static func tabsToRestore() -> [SavedTab]? {
```

Comment:
Since tabsState.archive is a global resource on disk and TabManager is not available when we want to launch the app without initializing anything I moved these to static to allow us to call it without a TabManager instance.